Dockerfile

```diff
@@ -14,64 +63,147 @@ COPY frontend/env/ ./frontend/env/
 ARG ENV=selfhosted
 RUN cd frontend && ENV=${ENV} npm ci --quiet --production
 
-# Copy the entire project - Webpack puts compiled assets into the Django folder
-COPY . .
+COPY frontend /app/frontend
+
+# * build-node-django [build-node]
+FROM build-node as build-node-django
+
+RUN mkdir /app/api
 ENV STATIC_ASSET_CDN_URL=/static/
 RUN cd frontend && npm run bundledjango
 
-# Step 2 - Build Python virtualenv
+# * build-node-selfhosted [build-node]
+FROM build-node as build-node-selfhosted
+
+RUN cd frontend && npm run bundle
+
+# * build-python
 FROM python:3.11 as build-python
 WORKDIR /app
 
 COPY api/pyproject.toml api/poetry.lock api/Makefile ./
-ARG POETRY_VIRTUALENVS_CREATE=false
-RUN make install-poetry
-ENV PATH="$PATH:/root/.local/bin"
+ENV POETRY_VIRTUALENVS_CREATE=false
+ENV POETRY_HOME=/usr/local
+RUN make install opts='--without dev'
 
-ARG GH_TOKEN
-RUN if [ -n "${GH_TOKEN}" ]; \
-  then echo "https://${GH_TOKEN}:@github.com" > ${HOME}/.git-credentials \
-  && git config --global credential.helper store; fi;
+# * build-python-private [build-python]
+FROM build-python AS build-python-private
 
-ARG POETRY_OPTS
-RUN make install-packages opts="${POETRY_OPTS}"
+# Authenticate git with token to install private packages
+RUN --mount=type=secret,id=github_private_cloud_token \
+  echo "https://$(cat /run/secrets/github_private_cloud_token):@github.com" > ${HOME}/.git-credentials && \
+  git config --global credential.helper store
 
-# Step 3 - Build Django Application
-FROM python:3.11-slim as application
-WORKDIR /app
+# Install SAML binary dependency
+RUN apt-get update && apt-get install -y xmlsec1
 
-# Install SAML dependency if required
-ARG SAML_INSTALLED="0"
-RUN if [ "${SAML_INSTALLED}" = "1" ]; then apt-get update && apt-get install -y xmlsec1; fi;
+# Install private Python dependencies
+RUN make install-packages opts='--without dev --with saml,auth-controller,ldap,workflows'
 
-# arm architecture platform builds need postgres drivers installing via apt
-ARG TARGETARCH
-RUN if [ "${TARGETARCH}" != "amd64" ]; then apt-get update && apt-get install -y libpq-dev && rm -rf /var/lib/apt/lists/*; fi;
+# Integrate private modules
+ARG SAML_REVISION
+ARG AUTH_CONTROLLER_REVISION
```

Comment:
Sorry for the collision but I think that (based on #4174) this is actually no longer needed.

Reply:
Removed.